spec/controllers/courses_controller_spec.rb

```diff
@@ -1,6 +1,85 @@
 describe CoursesController do
   let(:instructor) { create :instructor }
   let(:course) { instructor.course }
+
+  describe 'role switching methods' do
+    let(:end_user) { create :end_user }
+    let(:subject) do
+      post_as instructor, :switch_role, params: { 'id' => course.id, 'effective_user_login' => end_user.user_name }
+    end
+    describe '#switch_role' do
+      let(:second_instructor) { create :instructor }
```

Comment:
Set the course association for instructors/TAs/students you create here, to be explicit that they are part of the same course.

(Unless you are testing that they aren't. In fact, you should add tests checking that `:forbidden` status is returned if trying to switch role to a role in a different course.

Reply:
We currently do not have an option of returning the :forbidden or 403 status. Right now the routine is: 
- First we accquire the username. If it's blank we say username can't be blank and return a 404. 
- Then we try to find the user in the db + find their role in the course. If their role is nil we say login failed and return a 404
- Then we check whether they are logging in as themselves or another instructor in the course. Either case will be 422 but we'll give different messages
- Finally, we return 200 and do the stuff afterwards. 
I'm adding the tests for if they are switching to valid users not in the course - but let me know how you would like the status to be handled (I can either keep using 404 as their role will be nil as well, or I can modify the controller to use 403 with an extra if block).

Reply:
@TheTallJerry ah yes I see, a `:not_found` is totally fine. Thanks for the explanation. You can finalize the tests to check for `:not_found` (I guess just remove the parts you commented out.)